docs/language/informal/enhanced-type-promotion.md

```diff
@@ -0,0 +1,537 @@
+evaluates to true. We extend this to cover facts that are known when an
+expression evaluates to false.
+
+*   In a conditional expression of the form `e1 ? e2 : e3`, we know that if `e1`
+    shows that `v` has type T when `e1` is false, then `v` is known to be of
+    type T in `e3`:
+
+    ```dart
+    o is! String ? notHere : oPromotedToStringHere
+    ```
+
+*   In a logical boolean expression of the form `e1 || e2`, if `e1` shows `v`
+    has type T when `e1` is false, then `v` is known to be of type T in `e2`:
+
+    ```dart
+    o is! String || oPromotedToStringHere
+    ```
+
+    This is because short-circuiting means the right operand is only evaluated
+    when the left is false.
+
+*   In an if statement, if the condition shows that `v` has type T when the
+    condition is false, then `v` is known to be of type T in the else clause.
+
+    ```dart
+    if (o is! String) notHere; else oPromotedToStringHere;
+    ```
+
+### Loop Scopes
+
+Much like if statements, we extend C-style for and while loops to promote inside
+their bodies when the condition is true and the condition expression shows a `v`
+has type T:
+
+```dart
+while (o is String) oPromotedToStringHere;
+
+for (; o is String;;) oPromotedToStringHere;
+```
+
+### Scopes After Exits
+
+The last common case where users expect type promotion to work but currently
+doesn't is in code that can only be reached if a type test succeeds because
+earlier code will cause the code to exit, as in:
+
+```dart
+int stringLength(Object o) {
+  if (o is! String) throw "Not a string.";
+  return o.length; // Should be promoted here.
+}
+```
+
+We say a statement "cannot complete normally" if the statement is one of:
+
+*   A return statement
+*   A rethrow statement
+*   An expression statement whose expression is a throw expression
+*   An if statement with an else where neither statement can complete normally
+*   A block whose last statement cannot complete normally
+
+Then, given an if statement of the form `if (b) s1 else s2` that is contained in
+some block:
+
+*   If `b` shows that `v` has type T when `b` is true, and `s2` cannot complete
+    normally, then `v` has the type T in the statements after the if statement
+    in the immediately enclosing block. As in:
+
+    ```dart
+    int stringLength(Object o) {
+      if (o is String) /* whatever */; else throw "Not a string.";
+      return o.length; // Promoted here.
+    }
+    ```
+
+*   If `b` shows that `v` has type T when `b` is false, and `v` can be promoted,
+    and `s1` cannot complete normally, then `v` has the type T in the statements
+    after the if statement in the immediately enclosing block.
+
+    This is also allowed if the if statement has no else clause at all.
+
+*Optional: We could expand the definition of complete normally for a block to
+allow for returns or throws before the last statement as long as they are
+unconditionally executed.*
```

Comment:
We should. The following statements should not be treated differently, just because I had to add a block to make the code fit into 80 chars.

``` dart
if (e) return foo;
if (e) {
  return foo;
}
```

Adding curly braces is a style issue.

@johnniwinther